subprojects/build-operations/src/main/java/org/gradle/internal/operations/DefaultBuildOperationRunner.java

```diff
@@ -20,23 +20,26 @@
 import org.slf4j.LoggerFactory;
 
 import javax.annotation.Nullable;
-import javax.annotation.OverridingMethodsMustInvokeSuper;
-import java.io.ObjectStreamException;
-import java.util.concurrent.atomic.AtomicBoolean;
 
 public class DefaultBuildOperationRunner implements BuildOperationRunner {
     private static final Logger LOGGER = LoggerFactory.getLogger(DefaultBuildOperationRunner.class);
-    protected static final BuildOperationWorker<RunnableBuildOperation> RUNNABLE_BUILD_OPERATION_WORKER = new RunnableBuildOperationWorker();
+    private static final BuildOperationWorker<RunnableBuildOperation> RUNNABLE_BUILD_OPERATION_WORKER = new BuildOperationWorker<RunnableBuildOperation>() {
+        @Override
+        public void execute(RunnableBuildOperation buildOperation, BuildOperationContext context) throws Exception {
+            buildOperation.run(context);
+        }
+    };
 
-    protected final BuildOperationListener listener;
     private final TimeSupplier clock;
     private final BuildOperationIdFactory buildOperationIdFactory;
-    private final CurrentBuildOperationRef currentBuildOperationRef = CurrentBuildOperationRef.instance();
+    private final CurrentBuildOperationRef currentBuildOperationRef;
+    private final BuildOperationExecutionListenerFactory listenerFactory;
 
-    public DefaultBuildOperationRunner(BuildOperationListener listener, TimeSupplier clock, BuildOperationIdFactory buildOperationIdFactory) {
-        this.listener = listener;
+    public DefaultBuildOperationRunner(CurrentBuildOperationRef currentBuildOperationRef, TimeSupplier clock, BuildOperationIdFactory buildOperationIdFactory, BuildOperationExecutionListenerFactory listenerFactory) {
```

Comment:
❓ I'm not sure how I need to set the `CurrentBuildOperationRef` up on the Maven side.

Reply:
It is a weird singleton; you just pass `CurrentBuildOperationRef.instance()` like we do on this side I guess.

Reply:
But who is responsible of setting it up? Will that happen automagically?

Reply:
I don't really know. I don't understand why we have it a singleton the way it is, but that's what we have currently. We can ask @ldaley or convert it to a proper service in a separate PR.